lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
+        document_url,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        assignment = Assignment(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            document_url=document_url,
+            resource_link_id=resource_link_id,
+            ext_lti_assignment_id=ext_lti_assignment_id,
+        )
+        self._db.add(assignment)
+        return assignment
 
     def set_document_url(
-        self, tool_consumer_instance_guid, resource_link_id, document_url
+        self,
+        tool_consumer_instance_guid,
+        document_url,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
     ):
         """
-        Save the given document_url.
+        Save the given document_url in an existing assignment or create new one.
 
         Set the document_url for the assignment that matches
-        tool_consumer_instance_guid and resource_link_id. Any existing document
-        URL for this assignment will be overwritten.
-        """
-        assignment = self.get(tool_consumer_instance_guid, resource_link_id)
+        tool_consumer_instance_guid and resource_link_id/ext_lti_assignment_id
+        or create a new one if none exist on the DB.
 
-        if assignment:
+        Any existing document URL for this assignment will be overwritten.
+        """
+        try:
+            assignment = self.get(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
             assignment.document_url = document_url
-        else:
-            self._db.add(
-                Assignment(
-                    document_url=document_url,
-                    resource_link_id=resource_link_id,
-                    tool_consumer_instance_guid=tool_consumer_instance_guid,
-                )
+        except NoResultFound:
+            assignment = self.create(
+                document_url=document_url,
+                resource_link_id=resource_link_id,
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
             )
 
         # Clear the cache (@lru_cache) on self.get because we've changed the
         # contents of the DB. (Python's @lru_cache doesn't have a way to remove
         # just one key from the cache, you have to clear the entire cache.)
         self.get.cache_clear()
 
+        return assignment
+
+    def _merge_canvas_assignments(self, assignments):
+        # We stored a canvas file assignment before (storing its resource_link_id)
+        # we later configured it (storing its ext_lti_assignment_id) in a new row
+        # and now we are launching it, we want to merge those two assignments
+        old_assignment = (
+            assignments[0] if assignments[0].resource_link_id else assignments[1]
+        )
+        assignment = (
+            assignments[0] if assignments[0].ext_lti_assignment_id else assignments[1]
+        )
```

Comment:
Here you're implicitly assuming that one of the assignments has a `resource_link_id` and no `ext_lti_assignment_id` and the other has an `ext_lti_assignment_id` and no `resource_link_id`. I wonder if it might be nice to make that assumption explicit? Both because it might make the code more readable to explicitly state the assumption, and because it'll mean that we get notified if something's wrong.

This code also looks like it could assign the _same_ object to `old_assignment` and `assignment` in cases where the above assumption has failed. It might be good to just make it so that it can't possibly assign the two to the same object. Again both for readability and safety.

Maybe something like:

```python
if assignments[0].resource_link_id:
    old_assignment, new_assignment = assignments
else:
    new_assignment, old_assignment = assignments

assert not old_assignment.ext_lti_assignment_id
assert not new_assignment.resource_link_id
```

I think assertions might actually be fine and appropriate for this but could raise some other exception instead if you prefer.

Reply:
Yes good point, much more readable. Assertions are :+1: